eth_account/hdaccount/deterministic.py

```diff
@@ -0,0 +1,211 @@
+#   API, and no expectation is given for `xpub/xpriv` key derivation.
+from typing import (
+    Tuple,
+    Union,
+)
+
+from eth_utils import (
+    to_int,
+)
+
+from ._utils import (
+    SECP256K1_N,
+    ec_point,
+    hmac_sha512,
+)
+
+
+class Node(int):
+    TAG = ""  # No tag
+    OFFSET = 0x0  # No offset
+    """
+    Base node class
+    """
+    def __new__(cls, index):
+        obj = int.__new__(cls, index + cls.OFFSET)
+        obj.index = index
+        return obj
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.index})"
+
+    def __add__(self, other: int):
+        return self.__class__(self.index + other)
+
+    def serialize(self) -> bytes:
+        assert 0 <= self < 2**32
+        return self.to_bytes(4, byteorder="big")
+
+    def encode(self) -> str:
+        return str(self.index) + self.TAG
+
+    @staticmethod
+    def decode(node: str) -> Union["SoftNode", "HardNode"]:
+        if len(node) < 1:
+            raise ValueError("Cannot use empty string")
+        if node[-1] in ("'", "H"):
+            return HardNode(int(node[:-1]))
+        else:
+            return SoftNode(int(node))
+
+
+class SoftNode(Node):
+    """
+    Soft node (unhardened), where value = index
+    """
+    TAG = ""  # No tag
+    OFFSET = 0x0  # No offset
+
+
+class HardNode(Node):
+    """
+    Hard node, where value = index + BIP32_HARDENED_CONSTANT
+    """
+    TAG = "H"  # "H" (or "'") means hard node (but use "H" for clarity)
+    OFFSET = 0x80000000  # 2**31, BIP32 "Hardening constant"
+
+
+def derive_child_key(
+    parent_key: bytes,
+    parent_chain_code: bytes,
+    node: Node,
+) -> Tuple[bytes, bytes]:
+    """
+    From BIP32:
+
+    The function CKDpriv((k_par, c_par), i) → (k_i, c_i) computes a child extended
+    private key from the parent extended private key:
+
+    - Check whether i ≥ 2**31 (whether the child is a hardened key).
+      - If so (hardened child):
+        let I = HMAC-SHA512(Key = c_par, Data = 0x00 || ser_256(k_par) || ser_32(i)).
+        (Note: The 0x00 pads the private key to make it 33 bytes long.)
+      - If not (normal child):
+        let I = HMAC-SHA512(Key = c_par, Data = ser_P(point(k_par)) || ser_32(i)).
+    - Split I into two 32-byte sequences, I_L and I_R.
+    - The returned child key k_i is parse_256(I_L) + k_par (mod n).
+    - The returned chain code c_i is I_R.
+    - In case parse_256(I_L) ≥ n or k_i = 0, the resulting key is invalid,
+      and one should proceed with the next value for i.
+      (Note: this has probability lower than 1 in 2**127.)
+    """
+    assert len(parent_chain_code) == 32
+    if isinstance(node, HardNode):
+        # NOTE Empty byte is added to align to SoftNode case
+        assert len(parent_key) == 32  # Should be guaranteed here in return statment
+        child = hmac_sha512(parent_chain_code, b"\x00" + parent_key + node.serialize())
+
+    else:
```

Comment:
We have trended away from using *implicit* else clauses.  Am I correct that this case covers the `SoftNode` case?  If so I would advocate for an explicit check for `elif isinstance(node, SoftNode)` and then to have a final `else` clause that raises an exception like `TypeError` indicating that the `node` was not of a supported type.

Reply:
Will do